### ISSUE TITLE
TypeScript 5.7 Iteration Plan

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#59905]
This document outlines our focused tasks for TypeScript 5.7. It minimally indicates intent to investigate tasks or contribute to an implementation. Nothing is set in stone, but we will strive to complete these tasks in a reasonable timeframe.

Date           | Event
---------------|-------------------------
2024-09-09     | TypeScript 5.6 Release
2024-09-27     | Create 5.7 Beta (5.7.0) Build for Testing
2024-10-01     | **TypeScript 5.7 Beta Release**
2024-11-08     | Create 5.7 RC (5.7.1) Build for Testing
2024-11-12     | **TypeScript 5.7 RC Release**
2024-11-18     | Create 5.7 Final (5.7.2) Build for Testing
2024-11-21     | **TypeScript 5.7 Final Release** 🚀

# Compiler and Language

* [Control Flow Analysis for Lambdas Passed to `immediate` Parameters](https://github.com/microsoft/TypeScript/pull/58729)
* [Enforce Readonly Checks on Object Members](https://github.com/microsoft/TypeScript/pull/59326)
* [Checks for Never-Initialized Variables](https://github.com/microsoft/TypeScript/pull/55887)
* [Disallow Parameter Property References from Class Fields](https://github.com/microsoft/TypeScript/pull/59623)
* [Investigate Relating Values to Conditional Return Types](https://github.com/microsoft/TypeScript/issues/33912)
* [Investigate Relative Import Extension Rewrites](https://github.com/microsoft/TypeScript/pull/59767)
* [Investigate `/** @typeArguments/specialize */`](https://github.com/microsoft/TypeScript/pull/59666)
* [Investigate Support for Sourcemap v4](https://github.com/microsoft/TypeScript/issues/46695)
* [`lib.d.ts` Updates](https://github.com/microsoft/TypeScript/issues/59704)

# Editor and Language Service

* [Investigate Expandable Quick Info/Hover Verbosity](https://github.com/microsoft/TypeScript/issues/59029)
* [Consult Root Files Before Opening `composite` Projects](https://github.com/microsoft/TypeScript/pull/59688)
* [Ancestor Configuration File Searching](https://github.com/microsoft/TypeScript/issues/56959)
* [Completions for `package.json` Subpath `imports`](https://github.com/microsoft/TypeScript/pull/57718)
* [Improved Rename for Shorthand Properties/Destructuring](https://github.com/microsoft/TypeScript/issues/58447)
* [Support "Prepare Paste Edits" Command](https://github.com/microsoft/TypeScript/issues/59881)
* [Ship Import-on-Paste in Stable VS Code](https://github.com/microsoft/vscode/issues/30066)
* [Improved Detection of `node:` for Auto-Import Paths](https://github.com/microsoft/TypeScript/pull/59702)
* [Investigate File Drop Support in Editor](https://github.com/microsoft/TypeScript/issues/50170)
* [Investigate Improved Move to File Naming](https://github.com/microsoft/TypeScript/issues/46514)
* [Investigate LSP Support](https://github.com/microsoft/TypeScript/issues/39459)

# Performance

* [Path Mapping Optimizations](https://github.com/microsoft/TypeScript/pull/59048)
* [Investigate Enabling V8 Compile Caching in Node.js](https://github.com/microsoft/TypeScript/pull/59720)
* [Investigate and Experiment with Full Monomorphization](https://github.com/microsoft/TypeScript/pull/58928)

# Website and Docs

* [Simplify and Refactor Website for Faster Builds](https://github.com/microsoft/TypeScript-Website/issues/2730)
* Handbook Review
* Experiment with Example-Driven Learning Paths

# Infrastructure

* [Experiment with Reduced Repros on Weekly New Error Runs](https://github.com/microsoft/typescript-error-deltas/issues/164)
* [Consider `--isolatedDeclarations`](https://github.com/microsoft/TypeScript/pull/59635)
* [Use `ProjectService` when Running TS ESLint](https://github.com/microsoft/TypeScript/pull/59645)

**Comment by petamoriken** [comment]
I hope ES2024 target (#58573) will be added. It blocked by #59417.

**Comment by trevorade** [comment]
Are the dates for 5.7 still accurate or do they need an update?

I looked at the linked _Compiler and Language_ issues / PRs and most of them are not committed. The beta is scheduled 6 days from now. I'm unclear on what the typical timing is between commits and the Beta (or if all the intended commits are actually in the Beta)

**Comment by pociej** [comment]
Is 2024 target support planned to be included?

**Comment by HolgerJeromin** [comment]
@pociej 
#58573 was merged a few days ago, so yes this will be included.

**Comment by DanielRosenwasser** [comment]
@typescript-bot create release-5.7

**Comment by DanielRosenwasser** [comment]
@typescript-bot create release-5.7

**Comment by DanielRosenwasser** [comment]
@typescript-bot create release-5.7

**Comment by DanielRosenwasser** [comment]
Hey all, to be transparent, we have fallen behind a bit on the plans. The beta will be delayed by up to a week, though that doesn't mean we'll be bringing in everything that's still outstanding in the compiler + language. Thanks for the patience everyone!

**Comment by imbant** [comment]
Hi, I'm a language server developer. I am very curious about how to implement the Import-on-Paste function you mentioned above?
Here is my idea, is it feasible? (sorry but reading ts source code is a bit difficult for me)

1. Before pasting the text in the clipboard into the file, maintain a string of the pasted text in the memory
2. Compile the new code in this memory
3. If there is no syntax error, if an undefined symbol error is encountered in the line where the paste is located, try to do automatic completion at the location of the symbol
4. If there is an automatic completion item that completely matches the text of the symbol, complete the import statement corresponding to this item (the same as the behavior of the user manually selecting this completion item)
5. Repeat 2 to 5 until there is no such error, or there is no completely matching text
6. Overwrite the text string in the memory into the file to be pasted

**Comment by uhyo** [comment]
Hi, can I ask the current status of #33912 ?
I do understand that the iteration plan is not a commitment to complete it in 5.7, but I’d appreciate any information because I’m going to give a short talk about this exact feature. 😅 Particularly the info on what drove the addition to the plan would be great.

**Comment by uhyo** [comment]
👆 Solved by recently published Design Meeting Notes. Thank you!

**Comment by DanielRosenwasser** [comment]
Based on scheduling, we will be moving the date that 5.7 RC "snaps" to tomorrow, and release the RC next week. The stable release will still be released the week of the 21st.

**Comment by DanielRosenwasser** [comment]
@typescript-bot sync release-5.7

**Comment by DanielRosenwasser** [comment]
@typescript-bot bump release-5.7

**Comment by DanielRosenwasser** [comment]
As a heads up, it looks like a more realistic release date is tomorrow.

**Comment by DanielRosenwasser** [comment]
@typescript-bot bump release-5.7

**Comment by DanielRosenwasser** [comment]
Try harder robot.

@typescript-bot bump release-5.7

**Comment by trevorade** [comment]
Is there a timeline for the 5.7.3 release?

Also, so we can watch for it, approximate time when the 5.8 iteration plan will be created?

**Comment by newives** [comment]
Is there a timeline for the 5.7.3 release? 5.7.2 There are a lot of problems

**Comment by zedryas** [comment]
is they're a reason 5.8 is not announced like release where use to be ?

**Comment by HolgerJeromin** [comment]
Missing 5.7.2 (can be OK'ish) and 5.7.3 from nuget after 6 days :-(
https://www.nuget.org/packages/Microsoft.TypeScript.MSBuild
@DanielRosenwasser @weswigham
